wsmaster/che-core-api-factory/src/main/java/org/eclipse/che/api/factory/server/urlfactory/DefaultFactoryUrl.java

```diff
@@ -61,6 +67,38 @@ public String getBranch() {
     return null;
   }
 
+  @Override
+  public @Nullable String getCredentials() {
+    if (isNullOrEmpty(url)) {
+      return null;
+    }
+    Matcher matcher =
+        compile("https?://((?<username>[^:|@]+)?:?(?<password>[^@]+)?@)?.*").matcher(url);
+    String password = null;
+    String username = null;
+    try {
+      username = matcher.matches() ? matcher.group("username") : null;
+    } catch (IllegalArgumentException e) {
+      // no such group
+    }
+    try {
+      password = matcher.matches() ? matcher.group("password") : null;
+    } catch (IllegalArgumentException e) {
+      // no such group
+    }
+    if (!isNullOrEmpty(username) || !isNullOrEmpty(password)) {
+      return format(
+          "%s:%s",
+          isNullOrEmpty(username) ? "" : username, isNullOrEmpty(password) ? "" : password);
+    }
+    return null;
+  }
+
+  public <T extends DefaultFactoryUrl> T withUrl(String url) {
+    this.url = url;
+    return (T) this;
+  }
+
```

Comment:
can we possibly re-write it using Optional in order to avoid returning null and returning Optional.empty() / Optional.of(myString) ?

Reply:
done